usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -489,6 +491,394 @@ static CK_RV cleanse_attribute(TEMPLATE *template,
+        return rc;
+    }
+    rc = get_bool_attribute_by_type(pPrivateKeyTemplate, ulPrivateKeyAttributeCount,
+                                    CKA_IBM_ATTRBOUND, &abpriv);
+    if (rc != CKR_OK && rc != CKR_TEMPLATE_INCOMPLETE) {
+        TRACE_ERROR("Failed to retrieve CKA_IBM_ATTRBOUND attribute from private template. rc=0x%lx\n", rc);
+        return rc;
+    }
+    if (abpub != abpriv) {
+        TRACE_ERROR("Only one of the key pair is attribute bound.\n");
+        return CKR_TEMPLATE_INCONSISTENT;
+    }
+    return CKR_OK;
+}
+
+static CK_RV check_ab_kek_type(TEMPLATE *tmpl, CK_RV errval)
+{
+    CK_KEY_TYPE kektype;
+    CK_RV rc;
+
+    rc = template_attribute_get_ulong(tmpl, CKA_KEY_TYPE, &kektype);
+    if (rc == CKR_OK) {
+        if (kektype != CKK_RSA && !is_keytype_symmetric(kektype)) {
+            TRACE_ERROR("KEK key type %ld not supported for AB wrap/unwrap\n",
+                        kektype);
+            return errval;
+        }
+    }
+    return rc;
+}
+
+static CK_RV check_ab_attributes(CK_ATTRIBUTE *attrs,
+                                 CK_ULONG attrs_len,
+                                 CK_OBJECT_CLASS kc)
+{
+    CK_RV rc;
+    CK_BBOOL attrbound = FALSE, sensitive = FALSE;
+
+    if (kc != CKO_PUBLIC_KEY) {
+        rc = get_bool_attribute_by_type(attrs, attrs_len, CKA_IBM_ATTRBOUND,
+                                        &attrbound);
+        if (rc != CKR_OK && rc != CKR_TEMPLATE_INCOMPLETE)
+            return rc;
+        rc = get_bool_attribute_by_type(attrs, attrs_len, CKA_SENSITIVE,
+                                        &sensitive);
+        if (rc != CKR_OK && rc != CKR_TEMPLATE_INCOMPLETE)
+            return rc;
+        if (attrbound && !sensitive) {
+            TRACE_ERROR("Attribute bound key not sensitive!");
+            return CKR_TEMPLATE_INCONSISTENT;
+        }
+    }
+    return CKR_OK;
+}
+
+static CK_RV check_ab_wrap(STDLL_TokData_t * tokdata,
+                           CK_BYTE **sblob, size_t *sblob_len,
+                           OBJECT **sobj,
+                           OBJECT *key_obj, OBJECT *wrap_key_obj,
+                           CK_MECHANISM_PTR mech)
+{
+    CK_RV rc;
+    CK_BBOOL abkey, abwrapkey, absignkey, signsignkey;
+
+    *sobj = 0;
+    *sblob = 0;
+    *sblob_len = ~0;
+    rc = template_attribute_get_bool(key_obj->template, CKA_IBM_ATTRBOUND, &abkey);
+    if (rc == CKR_TEMPLATE_INCOMPLETE)
+        abkey = FALSE;
+    else if (rc != CKR_OK) {
+        TRACE_ERROR("Invalid CKA_IBM_ATTRBOUND attribute on key to wrap\n");
+        return rc;
+    }
+    rc = template_attribute_get_bool(wrap_key_obj->template, CKA_IBM_ATTRBOUND, &abwrapkey);
+    if (rc == CKR_TEMPLATE_INCOMPLETE)
+        abwrapkey = FALSE;
+    else if (rc != CKR_OK) {
+        TRACE_ERROR("Invalid CKA_IBM_ATTRBOUND attribute on wrapping key\n");
+        return rc;
+    }
+    if (abwrapkey) {
```

Comment:
If checking for (!abwrapkey) and exiting here, we would save an indentation

Reply:
I don't see that.  There is an `else` attached to this `if`.